x/epoch/types/hooks.go

```diff
@@ -0,0 +1,39 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type EpochHooks interface {
+	// the first block whose timestamp is after the duration is counted as the end of the epoch
+	AfterEpochEnd(ctx sdk.Context, epoch Epoch)
+	// new epoch is next block of epoch end block
+	BeforeEpochStart(ctx sdk.Context, epoch Epoch)
+}
+
+var _ EpochHooks = MultiEpochHooks{}
+
+// combine multiple gamm hooks, all hook functions are run in array sequence.
+type MultiEpochHooks []EpochHooks
+
+func NewMultiEpochHooks(hooks ...EpochHooks) MultiEpochHooks {
+	return hooks
+}
+
+// AfterEpochEnd is called when epoch is going to be ended, epochNumber is the number of epoch that is ending.
+func (h MultiEpochHooks) AfterEpochEnd(ctx sdk.Context, epoch Epoch) {
+	for i := range h {
+		cacheCtx, write := ctx.CacheContext()
```

Comment:
are the `AfterEpochEnd` and `BeforeEpochStart` supposed to break in any circumstance? i.e. when we do cacheContext and write() if the mint module's afterEpochEnd call fails somehow will the chain runs into a bad state indefinitely?

Reply:
\cc @codchen we were discussing this in his comment as well